File: src/main/java/core/procedures/vectors/VectorRef.java
package core.procedures.vectors;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Type;
import core.scm.MutableVector;

public final class VectorRef extends AFn {

  public VectorRef() {
    super(new FnArgsBuilder().min(2).max(2)
                             .mandatory(new Class[]{MutableVector.class, Type.ExactNonNegativeInteger.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "vector-ref";
  }

  @Override
  public Object apply2(Object arg1, Object arg2) {
    MutableVector vec = (MutableVector)arg1;
    Long pos = ((Number)arg2).longValue();
    if (pos >= vec.size()) {
      throw new IndexOutOfBoundsException(String.format("%s: value out of range: %s", getName(), pos));
    }
    return vec.get(pos.intValue());
  }
}

File: src/main/java/core/procedures/vectors/VectorRef.kt
package core.procedures.vectors

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Type
import core.scm.Vector

class VectorRef : AFn<Any?, Any?>(name = "vector-ref", isPure = true, arity = Exactly(2),
                      mandatoryArgsTypes = arrayOf(Vector::class.java, Type.ExactNonNegativeInteger::class.java)) {

    override operator fun invoke(arg1: Any?, arg2: Any?) = (arg1!! as Vector).invoke(arg2!! as Number)
}
